### PR TITLE
Add getAdvertistingPostByIdOrImportEntity support to mindful package

### PR DESCRIPTION
**Pull request by B77Mills**
- Add support for getAdvertisingPostByIdOrImportEntity to the api client.  
 - Centralize fragments, utilzs and route logic
 - Update marko-web mindful service with updated middleware calls for getAdvertisingPostByIdOrImportEntity
 
 http://www-wattagnet.dev.parameter1.com:9926/brand-insights/zoetis/vaccination-success-depends-on-crew-training/67094ddcc2bbee3e1b1af57d?pubid=644d5a33b225a200013c2dd2&utm_source=NativeX&utm_medium=banner&utm_campaign=67129a46b42e363e8d7c9ae3&utm_term=646792140ec6330001a96883&utm_content=
 
Screenshot 2024-11-04 at 10 53 32 AM

 
 
**start-server.js cal:**
 ```js
 const MindfulMarkoWebService = require('@parameter1/base-cms-mindful/marko-web/middleware/service');

const { namespace } = getAsObject(options, 'siteConfig.mindful');
app.use(MindfulMarkoWebService({ namespace }));

```
 
 
 **The following routes can be added withing the site repos with the following route calls:** 
 ```js
 // include at the top of routes:
 const advertisingPostByIdOrImportEntity = require('@parameter1/base-cms-mindful/marko-web/middleware/advertising-post-by-id-or-import-entity');
const advertisingPostTemplate = require('../templates/content/advertising-post');

// withing the routes list 
  getAdvertisingPostAsNativeStory(app, {
    route: '/brand-insights/:section/:slug/:id',
    tenant: 'watt',
    provider: 'native-x',
    template: advertisingPostTemplate,
  });
 ```
 
 **Optionally you can also set a fragment to override the following:**
 ```js
 
 const advertisingPostById = require('../../graphql/fragments/advertising-post-by-id');

 // withing the routes list 
  advertisingPostByIdOrImportEntity(app, {
    route: '/brand-insights/:section/:slug/:id',
    tenant: 'watt',
    provider: 'native-x',
    template: advertisingPostTemplate,
    fragment: advertisingPostById
  });
  ```
**Example Fragment**
 ```gql
 
 fragment AdvertisingPostById on AdvertisingPost {
  _id
  title {
    default
  }
  teaser
  body
  url
  publishedDay
  statusEdge {
    node {
      _id
      label
    }
  }
  featuredImageEdge {
    _id
    node {
      _id
      src {
        url
        settings {
          fpX
          fpY
        }
      }
    }
  }
  companyEdge {
    _id
    node {
      _id
      name {
        default
      }
      logoEdge {
        _id
        node {
          _id
          src {
            url
            settings {
              fpX
              fpY
            }
          }
        }
      }
    }
  }
}
```

### CODE REVIEW

File: packages/mindful/api-client.js [hunk 1]
```diff
@@ -51,6 +51,41 @@ class MindfulApiClient {
...
+   * @param {string} args.provider
+   * @param {string} args.tenant
+   * @param {import("graphql").DocumentNode|string} fragment
+   */
+  async getAdvertisingPostByIdOrImportEntity(
+    {
+      _id,
+      provider,
+      tenant,
+      type = 'post',
```
Comment (zarathustra323) [inline packages/mindful/api-client.js]: `type` shouldn't have a default value set

File: packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+const { asyncRoute } = require('@parameter1/base-cms-utils');
```
Comment (zarathustra323) [inline packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js]: There only needs to be one marko web middleware... and that's this one. Rename this to `advertising-post-as-native-story.js` and then delete the `advertising-post-by-id.js`

File: packages/mindful/marko-web/middleware/advertising-post-by-id.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const { asyncRoute } = require('@parameter1/base-cms-utils');
+const { get } = require('@parameter1/base-cms-object-path');
```
Comment (zarathustra323) [inline packages/mindful/marko-web/middleware/advertising-post-by-id.js]: This file can be removed

File: packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  template,
+  tenant,
+  provider,
+  fragment = defaultFragment,
+}) => {
+  app.get(route, asyncRoute(async (req, res) => {
+    const { service } = req.mindful;
+    const { id: _id } = req.params;
+    const { namespace } = tenant;
+    const result = await service.getAdvertisingPostByIdOrImportEntity({
```
Comment (zarathustra323) [inline packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js]: Rename the service method to `getAdvertisingPostAsNativeStory` and move the conversion to that method... see below.

In other words this should become
```js
const story = service.getAdvertisingPostAsNativeStory({ _id, preview: Boolean(req.query.preview), tenant });
res.marko(template, { story });
```

File: packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+      _id,
+      provider,
+      tenant: namespace,
```
Comment (zarathustra323) [inline packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js]: The `namespace` is not the tenant in this case... the tenant is the `native-x` tenant key, e.g. `watt`, not `watt/default`

File: packages/mindful/marko-web/service.js [hunk 6]
```diff
@@ -20,5 +20,26 @@ class MindfulMarkoWebService {
...
     return post;
   }
+
+  /**
+   * Get post by Id
+   *
+   * @param {string} externalId
+   * @returns {Promise}
+   */
+  async getAdvertisingPostByIdOrImportEntity({
```
Comment (zarathustra323) [inline packages/mindful/marko-web/service.js]: Change this to `getAdvertisingPostAsNativeStory` with `_id`, `tenant` and `preview` as args. The provider can be hardset to `native-x` and the `type` hard set to `story`.

Use the `convertAdvertisingPostToNativeStory` directly in this method.

File: packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    }, fragment);
+    const story = service.convertAdvertisingPostToNativeStory({
```
Comment (zarathustra323) [inline packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js]: this middleware should also return a 404 page ithe story cannot be found.

File: packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      advertisingPost: getAsObject(result, 'data.advertisingPostByIdOrImportEntity'),
+      preview: Boolean(req.query.preview),
+    });
+    res.marko(template, { story });
```
Comment (zarathustra323) [inline packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js]: There used to be handling for redirecting the page is the slug is incorrectly entered or if it changes... but I don't see that happening here?

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

File: packages/mindful/api-client.js [hunk 9]
```diff
@@ -51,6 +51,43 @@ class MindfulApiClient {
...
+   * @param {string} args._id
+   * @param {string} args.provider
+   * @param {string} args.tenant
+   * @param {import("graphql").DocumentNode|string} fragment
+   */
+  async getAdvertisingPostByIdOrImportEntity(
+    {
+      _id,
+      provider,
+      tenant = 'native-x',
```
Comment (zarathustra323) [inline packages/mindful/api-client.js]: Remove default values from `tenant` and `type` params

File: packages/mindful/marko-web/middleware/get-advertising-post-as-native-story.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+const createError = require('http-errors');
+const { asyncRoute } = require('@parameter1/base-cms-utils');
+const defaultFragment = require('../../graphql/fragments/advertising-post-by-id');
+
+module.exports = (app, {
+  route,
+  template,
+  tenant,
+  provider,
```
Comment (zarathustra323) [inline packages/mindful/marko-web/middleware/get-advertising-post-as-native-story.js]: Do not pass the `provider` param

File: packages/mindful/marko-web/middleware/get-advertising-post-as-native-story.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+      return;
+    }
+
+    const { service } = req.mindful;
+    const { id: _id } = req.params;
+    const story = await service.getAdvertisingPostByIdOrImportEntity({
+      _id,
+      preview: Boolean(req.query.preview),
+      tenant,
+      provider,
```
Comment (zarathustra323) [inline packages/mindful/marko-web/middleware/get-advertising-post-as-native-story.js]: Do not pass the `provider` param

File: packages/mindful/marko-web/service.js [hunk 12]
```diff
@@ -16,9 +17,22 @@ class MindfulMarkoWebService {
    * @param {string} externalId
    * @returns {Promise}
    */
-  async getAdvertisingPostById({ _id }, fragment) {
-    const post = await this.client.getAdvertisingPostById({ _id }, fragment);
-    return post;
+  async getAdvertisingPostAsNativeStory({
+    _id,
+    preview,
```
Comment (zarathustra323) [inline packages/mindful/marko-web/service.js:22]: Remove the `provider` and `type` params.. only allow `tenant`

File: packages/mindful/marko-web/service.js [hunk 13]
```diff
@@ -16,9 +17,22 @@ class MindfulMarkoWebService {
...
+    provider,
+    tenant,
+    type = 'story',
+  }, fragment) {
+    const response = await this.client.getAdvertisingPostByIdOrImportEntity({
+      _id,
+      preview,
+      provider,
```
Comment (zarathustra323) [inline packages/mindful/marko-web/service.js]: Hard set provider and type to:
```js
provider: 'native-x',
type: 'story',
```

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

**Review by zarathustra323 — APPROVED** [review]

**Comment by B77Mills** [comment]
[Add monorail/mindful package to replace with native story middleware with Mindful equivalent](https://trello.com/c/zU4rTgw1/547-add-monorail-mindful-package-to-replace-with-native-story-middleware-with-mindful-equivalent)

**Pull request merged**